docs/architecture/overview.md

```diff
@@ -0,0 +1,152 @@
+# Overview
+
+The components that make up a running test are simple from the outside.
+
+The idea behind system tests is that we can share the tests for a given feature across implementations.
+
+Enabling a feature within system tests might go like this:
+ - Add the enabling environment variable to the shared docker-compose
+ - Add a test asserting the environment variable affected traces as expected
+ - Build and run for each language to verify the tests run and behave as expected
+ - Create the pull request in system-tests
+
+However, there are many scenarios where a test may not be so simple to implement.
+
+This document aims to give a working understanding of the parts of system-tests, and how to troubleshoot them.
+
+## What are the components of a running test?
+
+When the system tests are executing, there are several main containers of concern.
+ - [Tests Container](TODO)
+   - Responsible for running the actual tests, sending traffic, and asserting results
+ - [Application Container](#structure-of-the-application-container)
+   - Swappable webapp language module that must meet an interface
+ - [Application Proxy](TODO)
+   - Mechanism to inspect payloads from the datadog libraries
+ - [Agent](TODO)
+   - Basic Datadog agent image
+ - [Agent Proxy](TODO)
+   - Mechanism to inspect payloads from the Agent to the Backend
+
+```mermaid
+flowchart TD
+    TESTS[Tests Container] -->|Send Requests| APP
+    APP[Application Container] --> APPPROXY
+    APPPROXY[Application Proxy] --> AGENT
+    AGENT[Agent Container] --> AGENTPROXY
+    AGENTPROXY[Agent Proxy] -->|remote request| BACKEND
+    BACKEND[Datadog]
+```
+
+The tests send requests directly to the application.
+The tests then wait on the results, which are available as the logs are collected from [mitmproxy](TODO) dumps.
+
+## What are system-tests bad for?
+
+ - Combinatorial style tests
+ - Tests that require specific versions of runtimes, libraries, or operating systems
```

Comment:
Even if it's not as easy as adding a test case of an existing test, it's kind of doable as long as it runs on docker. For instance, ruby engineer added lot of rails/sinatra version with a reasonable effort. 

I would say that even if it's not build to crawl a version matrix, it's made to be easy enough to add new frameworks/runtime, or relevants version of them as long as it runs on docker.

Reply:
Clarified this a bit